### ISSUE TITLE
Better diagnostics for unlinked projects

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#8622]
Problem: when opening a project with unusual struture (Cargo.toml buried deep in the source tree) rust-analyzer doesn't work, *without a clear error message*.

At the moment, when opening the project, we to detect rust project structure we either use `likedProjects` config,  or, if that is absent, look for Cargo.tomls in level one subdirs. So, we fail to discover deeply nested projects.

Soln1: travers all folders, and not only the top level. This is a bad idea, as this is an unbounded amount of work.

Soln2: lazily load Cargo.projects. That is, when a user opens an .rs file which does not belong to any known project, traverse the directory upwards looking for Cargo.toml. This seems like a best solution UX wise on the first glance. There's a big problem though -- non-determinism. If there are two projects, A and B, and A uses B, then opening B first and invoking find useages won't find usages in A, because we don't know about A yet!

Soln3, what we should do: start like soln2, but, instead of magically loading the file, show a popup "add this file to a list of linked projects?". That way, discovery results are persistent, and user always gets the same workspace.

**Comment by flodiebold** [comment]
> If there are two projects, A and B, and A uses B, then opening B first and invoking find useages won't find usages in A, because we don't know about A yet!

If A and B aren't in a common Cargo workspace and A just uses B through a path dependency, would we actually find the usages? My guess would be that we see them as completely separate crate graphs?

**Comment by matklad** [comment]
If we have an entry for A in the linked projects, then yes, we'll find usages (or at least, will be able to find them in theory, depending on how our source2def picks B).

**Comment by jonas-schievink** [comment]
> Soln1: travers all folders, and not only the top level. This is a bad idea, as this is an unbounded amount of work.

It seems like this is a problem for big workspaces, but I've often used projects where just looking into every top-level directory would be enough to discover all Cargo.tomls needed. So perhaps just descending a single level would be another possible solution here.

It would also be interesting to look at how other IDEs do this.

**Comment by matklad** [comment]
>It seems like this is a problem for big workspaces, but I've often used projects where just looking into every top-level directory would be enough to discover all Cargo.tomls needed. So perhaps just descending a single level would be another possible solution here.

That's exactly what we are doing! ;)

>It would also be interesting to look at how other IDEs do this.

IntelliJ Rust does 3, with a caveat that the list of liked projects is managed internally and is not exposed via user-editable config file. That is, instead of JSON, there's a GUI for the list of associated projects

![image](https://user-images.githubusercontent.com/1711539/115717288-b30d4780-a382-11eb-873a-07d703f9b730.png)

**Comment by flodiebold** [comment]
The problem with changing the configs currently is that it depends on the client, unless we finally introduce `rust-analyzer.toml` :grimacing:

**Comment by matklad** [comment]
Yeah, that's true... I wonder if we should add a "change config" LSP extension?

**Comment by flodiebold** [comment]
See #8631, we might want to lazily load non-toplevel Cargo projects in any case...

**Comment by flodiebold** [comment]
This is also related to #6388, I think -- if we're able to load Cargo projects when you open a file from them, the next step would be to create a fake crate for free-standing files when you open them...

**Comment by flodiebold** [comment]
Btw, we also have the "changing configs" use case for cargo features (toggling them on/off)...

**Comment by matklad** [comment]
Uhh 
![](https://intellij-rust.github.io/assets/posts/changelog-146/intention-cargo-feature.gif)

**Comment by bugadani** [comment]
Even if linkedProject points to the workspace root Cargo.toml, vscode may not think my source files are linked:
![image](https://user-images.githubusercontent.com/977627/134392714-bd38d307-f3b9-4fbe-80ba-b373466dc54f.png)

repro: https://github.com/bugadani/rust-esp32-example

**Comment by flodiebold** [comment]
@bugadani you need to configure linkedProjects like
```json
{
    "rust-analyzer.linkedProjects": ["components/rustlib/Cargo.toml"]
}
```
, not with an inline rust-project.json.

**Comment by bugadani** [comment]
Ah, shoot. Thanks a lot and sorry for the noise. For whatever reason I found the "Non-Cargo Based Projects" section in the docs, but did not search back for "linkedProjects" and missed that, badly. This means I also misunderstood https://github.com/rust-analyzer/rust-analyzer/issues/7268#issuecomment-925118794 Oh well. Thanks again.

**Comment by nickwesselman** [comment]
We have a standard project structure that results in Rust packages two levels deep, and thus not working with Rust Analyzer.  `linkedProjects` is a workaround but requires configuring each package individually. Would support for wildcards in `linkedProjects` be an option?

```
{
    "rust-analyzer.linkedProjects": [
        "extensions/*/Cargo.toml"
    ]
}
```

**Comment by stephenandary** [comment]
> We have a standard project structure that results in Rust packages two levels deep, and thus not working with Rust Analyzer. `linkedProjects` is a workaround but requires configuring each package individually. Would support for wildcards in `linkedProjects` be an option?
> 
> ```
> {
>     "rust-analyzer.linkedProjects": [
>         "extensions/*/Cargo.toml"
>     ]
> }
> ```

I was also hoping for something like this, or being able to define the depth of search like you can for git repos.

**Comment by stephenandary** [comment]
Hello :), wondering if there are any updates on being able to use wildcards with linkedProjects?

**Comment by fmhr** [comment]
vscode does not support wildcards, so this is what we need.
https://github.com/Microsoft/vscode/issues/58698
